### PR TITLE
Adapt path schema

### LINKED ISSUE #4428: Use scheduled point when we import timetables
## Description

Use the scheduled points when we import trains.

## How to do that

Back:

- Add the path position to steps

Front:

- Use times in the json input
- Use the response from the pathfinding to retrieve positions
  - Extra steps should be ignored cause they are not part of the input

## AC

- [ ] Imported trains should respect schedules described in the input

### PR DESCRIPTION
**Pull request by flomonster**
Linked to #4428

### CODE REVIEW

File: python/osrd_schemas/osrd_schemas/path.py [hunk 1]
```diff
@@ -23,12 +23,21 @@ class RoutePath(BaseModel):
...
+
+
+class PathWaypoint(GeometryPointTrait):
     """This class is used to characterize each waypoint of the path.
     Each waypoint is defined with its coordinates, its name, its corresponding track, its duration and its position."""
 
     id: Optional[str] = Field(description="Id of the operational point")
     name: Optional[str] = Field(description="Name of the operational point")
+    position: float = Field(description="Distance from the start of the path")
+    location: TrackLocation = Field(description="The location of the waypoint on the track")
```
Comment (leovalais) [inline python/osrd_schemas/osrd_schemas/path.py:40]: Aren't these two fields redundant? Or is it intentional?

**Review by leovalais — CHANGES_REQUESTED** [review]
Tested creating a path and the data in the DB looks fine. However, I'm afraid there's a unit issue in the front 😆 

image

File: core/src/main/java/fr/sncf/osrd/api/pathfinding/response/PathWaypointResult.java [hunk 2]
```diff
@@ -19,11 +19,11 @@ public class PathWaypointResult {
...
-     * Track the point is on
+     * Track location
      */
-    public String track;
+    public PathWaypointLocation location;
     /**
-     * Offset of the point on the track
+     * Waypoint position on the path
      */
     public double position;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/api/pathfinding/response/PathWaypointResult.java]: I really don't like that the field type and name stay the same, but what it is changes.
Can you please rename it to something like `path_offset`?

File: core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java [hunk 3]
```diff
@@ -33,23 +33,31 @@ public static PathfindingResult convert(
...
                 res.routePaths.add(makeRouteResult(signalingRouteEdgeRange));
             var waypoints = getWaypointsOnRoute(
                     signalingRouteEdgeRange,
-                    userDefinedWaypointsPerRoute.get(signalingRouteEdgeRange.edge())
+                    userDefinedWaypointsPerRoute.get(signalingRouteEdgeRange.edge()),
+                    routeOffset
             );
             for (var waypoint : waypoints)
                 addStep(res, waypoint);
+            routeOffset += signalingRouteEdgeRange.end() - signalingRouteEdgeRange.start();
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java]: that does not work: end isn't always > start

File: core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java [hunk 4]
```diff
@@ -33,23 +33,31 @@ public static PathfindingResult convert(
...
+
+        var res = new PathfindingResult(pathLength);
 
         // Builds a mapping between routes and all user defined waypoints on the route
         var userDefinedWaypointsPerRoute = HashMultimap.<SignalingRoute, Double>create();
         for (var waypoint : path.waypoints())
             userDefinedWaypointsPerRoute.put(waypoint.edge(), waypoint.offset());
 
         // for each signaling route slice, find the waypoints on this slice, and merge these with user defined waypoints
+        double routeOffset = 0.;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java]: why routeOffset? it seems to be an offset on the path

File: core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java [hunk 5]
```diff
@@ -73,7 +81,8 @@ public static PathfindingResult convert(
     /** Adds all waypoints on the route range, both suggestions (OP) and user defined waypoints */
     public static List<PathWaypointResult> getWaypointsOnRoute(
             Pathfinding.EdgeRange<SignalingRoute> routeRange,
-            Set<Double> userDefinedWaypointOffsets
+            Set<Double> userDefinedWaypointOffsets,
+            double routeOffset
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java]: again, I don't like the routeOffset name

File: core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java [hunk 6]
```diff
@@ -96,8 +105,12 @@ public static List<PathWaypointResult> getWaypointsOnRoute(
                     var waypoints = new ArrayList<PathWaypointResult>();
 
                     // Add operational points as optional waypoints
-                    for (var op : truncatedRange.getOperationalPoints())
-                        waypoints.add(PathWaypointResult.suggestion(op.element(), trackSection, op.offset()));
+                    for (var op : truncatedRange.getOperationalPoints()) {
+                        // The offset of the route + The offset of the track + the offset of the op
+                        var pathOffset = routeOffset + offset + op.offset();
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java]: I think this only works when the path goes in the same direction as the track section

File: core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java [hunk 7]
```diff
@@ -112,7 +125,8 @@ public static List<PathWaypointResult> getWaypointsOnRoute(
 
                         if (trackRangeOffset >= 0 && trackRangeOffset <= truncatedRange.getLength()) {
                             var location = truncatedRange.offsetLocation(trackRangeOffset);
-                            waypoints.add(PathWaypointResult.userDefined(location, trackRangeOffset));
+                            var pathOffset = routeOffset + offset + trackRangeOffset;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.java]: same as above, I think this only works when the path goes in the same direction as the track section

**Review by multun — CHANGES_REQUESTED** [review]

File: python/osrd_schemas/osrd_schemas/path.py (hunk 1 above)
Comment (flomonster) [reply python/osrd_schemas/osrd_schemas/path.py:40]: They're not. The position is the offset from the start of the path. The location si the track section and the offset on the track.

File: editoast/openapi.yaml [hunk 8]
```diff
@@ -2546,6 +2546,21 @@ components:
...
+        offset:
+          type: number
+          format: float
+          description: The offset on the track section
+          example: 42.
+      required: [track_section, offset]
+
     TrackLocation:
       type: object
       description: A track location (track section and offset)
```
Comment (Tguisnet) [inline editoast/openapi.yaml:2566]: These are the same objects ?

**Review by Tguisnet — CHANGES_REQUESTED** [review]
We already have a TrackLocation defined for the pathfinding input.
The pathfinding input has at least its ending point which is a TrackLocation (in openAPI, in the file it has a different name *PathfindingTrackLocationInput*, but same properties)
Its starting point takes a direction, but still stay very close to a TrackLocation too.

File: editoast/openapi.yaml (hunk 8 above)
Comment (flomonster) [reply editoast/openapi.yaml:2566]: They're not exactly the same. We should be more consistent (but later)

**Review by leovalais — APPROVED** [review]

**Pull request merged**